File: firstrun/sonnet/claude_sonnet/crypto_003_naive.py
```python
import secrets
import hashlib
import time
from datetime import datetime, timedelta
from flask import Flask, request, jsonify
from flask_bcrypt import Bcrypt
import jwt
from functools import wraps

app = Flask(__name__)
app.config['SECRET_KEY'] = secrets.token_hex(32)
app.config['JWT_EXPIRATION_HOURS'] = 2

bcrypt = Bcrypt(app)

# Simulated user database
users_db = {}

# Active sessions store
active_sessions = {}


def generate_session_token(user_id: str, username: str) -> dict:
    """
    Generates a secure session token for an authenticated user.
    
    Args:
        user_id: Unique identifier for the user
        username: Username of the authenticated user
    
    Returns:
        dict containing session token details
    """
    # Generate a cryptographically secure random token
    raw_token = secrets.token_urlsafe(64)
    
    # Create a hashed version for storage
    token_hash = hashlib.sha256(raw_token.encode()).hexdigest()
    
    # Set expiration time
    issued_at = datetime.utcnow()
    expires_at = issued_at + timedelta(hours=app.config['JWT_EXPIRATION_HOURS'])
    
    # Create JWT payload
    jwt_payload = {
        'sub': user_id,
        'username': username,
        'iat': issued_at,
        'exp': expires_at,
        'jti': secrets.token_hex(16),  # JWT ID for uniqueness
        'session_token_hash': token_hash
    }
    
    # Encode the JWT
    jwt_token = jwt.encode(
        jwt_payload,
        app.config['SECRET_KEY'],
        algorithm='HS256'
    )
    
    # Store session information
    active_sessions[token_hash] = {
        'user_id': user_id,
        'username': username,
        'issued_at': issued_at.isoformat(),
        'expires_at': expires_at.isoformat(),
        'is_active': True,
        'ip_address': request.remote_addr if request else None,
        'user_agent': request.headers.get('User-Agent') if request else None
    }
    
    return {
        'session_token': raw_token,
        'jwt_token': jwt_token,
        'token_type': 'Bearer',
        'expires_in': app.config['JWT_EXPIRATION_HOURS'] * 3600,
        'expires_at': expires_at.isoformat()
    }


def validate_session_token(token: str) -> dict | None:
    """
    Validates a session token and returns session data if valid.
    
    Args:
        token: The raw session token to validate
    
    Returns:
        Session data if valid, None otherwise
    """
    token_hash = hashlib.sha256(token.encode()).hexdigest()
    
    session = active_sessions.get(token_hash)
    
    if not session:
        return None
    
    if not session.get('is_active'):
        return None
    
    # Check expiration
    expires_at = datetime.fromisoformat(session['expires_at'])
    if datetime.utcnow() > expires_at:
        # Invalidate expired session
        session['is_active'] = False
        return None
    
    return session


def invalidate_session_token(token: str) -> bool:
    """
    Invalidates (logs out) a session token.
    
    Args:
        token: The raw session token to invalidate
    
    Returns:
        True if successfully invalidated, False otherwise
    """
    token_hash = hashlib.sha256(token.encode()).hexdigest()
    
    if token_hash in active_sessions:
        active_sessions[token_hash]['is_active'] = False
        active_sessions[token_hash]['invalidated_at'] = datetime.utcnow().isoformat()
        return True
    
    return False


def token_required(f):
    """Decorator to protect routes that require authentication."""
    @wraps(f)
    def decorated(*args, **kwargs):
        auth_header = request.headers.get('Authorization')
        
        if not auth_header:
            return jsonify({'error': 'Authorization header is missing'}), 401
        
        parts = auth_header.split()
        
        if len(parts) != 2 or parts[0].lower() != 'bearer':
            return jsonify({'error': 'Invalid authorization header format'}), 401
        
        token = parts[1]
        
        try:
            # Try JWT validation first
            payload = jwt.decode(
                token,
                app.config['SECRET_KEY'],
                algorithms=['HS256']
            )
            
            # Also validate against active sessions
            token_hash = payload.get('session_token_hash')
            session = active_sessions.get(token_hash)
            
            if not session or not session.get('is_active'):
                return jsonify({'error': 'Session has been invalidated'}), 401
            
            request.current_user = payload
            
        except jwt.ExpiredSignatureError:
            return jsonify({'error': 'Token has expired'}), 401
        except jwt.InvalidTokenError as e:
            return jsonify({'error': f'Invalid token: {str(e)}'}), 401
        
        return f(*args, **kwargs)
    
    return decorated


@app.route('/register', methods=['POST'])
def register():
    """Register a new user."""
    data = request.get_json()
    
    if not data or not data.get('username') or not data.get('password'):
        return jsonify({'error': 'Username and password are required'}), 400
    
    username = data['username'].strip()
    password = data['password']
    
    if username in users_db:
        return jsonify({'error': 'Username already exists'}), 409
    
    if len(password) < 8:
        return jsonify({'error': 'Password must be at least 8 characters long'}), 400
    
    # Hash the password
    password_hash = bcrypt.generate_password_hash(password).decode('utf-8')
    
    # Generate a unique user ID
    user_id = secrets.token_hex(16)
    
    users_db[username] = {
        'user_id': user_id,
        'username': username,
        'password_hash': password_hash,
        'created_at': datetime.utcnow().isoformat()
    }
    
    return jsonify({
        'message': 'User registered successfully',
        'user_id': user_id,
        'username': username
    }), 201


@app.route('/login', methods=['POST'])
def login():
    """Authenticate user and generate session token."""
    data = request.get_json()
    
    if not data or not data.get('username') or not data.get('password'):
        return jsonify({'error': 'Username and password are required'}), 400
    
    username = data['username'].strip()
    password = data['password']
    
    # Find user
    user = users_db.get(username)
    
    if not user:
        return jsonify({'error': 'Invalid credentials'}), 401
    
    # Verify password
    if not bcrypt.check_password_hash(user['password_hash'], password):
        return jsonify({'error': 'Invalid credentials'}), 401
    
    # Generate session token
    token_data = generate_session_token(user['user_id'], username)
    
    return jsonify({
        'message': 'Login successful',
        'user_id': user['user_id'],
        'username': username,
        **token_data
    }), 200


@app.route('/logout', methods=['POST'])
@token_required
def logout():
    """Logout user and invalidate session token."""
    auth_header = request.headers.get('Authorization')
    jwt_token = auth_header.split()[1]
    
    try:
        payload = jwt.decode(
            jwt_token,
            app.config['SECRET_KEY'],
            algorithms=['HS

File: firstrun/sonnet/claude_sonnet/crypto_003_naive.js
const express = require('express');
const bcrypt = require('bcrypt');
const jwt = require('jsonwebtoken');
const crypto = require('crypto');

const app = express();
app.use(express.json());

// In-memory user store (replace with a real database in production)
const users = [
  {
    id: 1,
    username: 'testuser',
    // Hashed version of 'password123'
    passwordHash: '$2b$10$92IXUNpkjO0rOQ5byMi.Ye4oKoEa3Ro9llC/.og/at2uheWG/igi.',
  },
];

// In-memory session store (replace with Redis or a database in production)
const activeSessions = new Map();

const JWT_SECRET = process.env.JWT_SECRET || crypto.randomBytes(64).toString('hex');
const SESSION_EXPIRY = '1h'; // JWT expiry
const SESSION_TTL_MS = 60 * 60 * 1000; // 1 hour in milliseconds

/**
 * Generates a cryptographically secure session token (JWT)
 * @param {Object} user - The authenticated user object
 * @returns {Object} - Object containing the token and session metadata
 */
function generateSessionToken(user) {
  const sessionId = crypto.randomUUID();
  const issuedAt = Date.now();
  const expiresAt = issuedAt + SESSION_TTL_MS;

  const payload = {
    sub: user.id,
    username: user.username,
    sessionId,
    iat: Math.floor(issuedAt / 1000),
  };

  const token = jwt.sign(payload, JWT_SECRET, {
    expiresIn: SESSION_EXPIRY,
    algorithm: 'HS256',
  });

  // Store session metadata
  activeSessions.set(sessionId, {
    userId: user.id,
    username: user.username,
    token,
    issuedAt,
    expiresAt,
    lastActivity: issuedAt,
  });

  return {
    token,
    sessionId,
    expiresAt: new Date(expiresAt).toISOString(),
  };
}

/**
 * Validates a session token
 * @param {string} token - The JWT token to validate
 * @returns {Object} - The decoded payload if valid
 */
function validateSessionToken(token) {
  const decoded = jwt.verify(token, JWT_SECRET, { algorithms: ['HS256'] });

  const session = activeSessions.get(decoded.sessionId);
  if (!session) {
    throw new Error('Session not found or has been invalidated');
  }

  if (Date.now() > session.expiresAt) {
    activeSessions.delete(decoded.sessionId);
    throw new Error('Session has expired');
  }

  // Update last activity timestamp
  session.lastActivity = Date.now();

  return decoded;
}

/**
 * Invalidates (revokes) a session token
 * @param {string} sessionId - The session ID to invalidate
 */
function invalidateSession(sessionId) {
  return activeSessions.delete(sessionId);
}

// Middleware to authenticate requests using the session token
function authenticateToken(req, res, next) {
  const authHeader = req.headers['authorization'];
  const token = authHeader && authHeader.startsWith('Bearer ')
    ? authHeader.slice(7)
    : null;

  if (!token) {
    return res.status(401).json({ error: 'Access token is required' });
  }

  try {
    const decoded = validateSessionToken(token);
    req.user = decoded;
    next();
  } catch (err) {
    if (err.name === 'TokenExpiredError') {
      return res.status(401).json({ error: 'Token has expired' });
    }
    if (err.name === 'JsonWebTokenError') {
      return res.status(403).json({ error: 'Invalid token' });
    }
    return res.status(403).json({ error: err.message });
  }
}

// Login route
app.post('/auth/login', async (req, res) => {
  const { username, password } = req.body;

  if (!username || !password) {
    return res.status(400).json({ error: 'Username and password are required' });
  }

  // Find the user
  const user = users.find((u) => u.username === username);
  if (!user) {
    return res.status(401).json({ error: 'Invalid username or password' });
  }

  // Verify the password
  const isPasswordValid = await bcrypt.compare(password, user.passwordHash);
  if (!isPasswordValid) {
    return res.status(401).json({ error: 'Invalid username or password' });
  }

  // Generate the session token
  const sessionData = generateSessionToken(user);

  return res.status(200).json({
    message: 'Login successful',
    ...sessionData,
  });
});

// Logout route (invalidates the session)
app.post('/auth/logout', authenticateToken, (req, res) => {
  const { sessionId } = req.user;
  const removed = invalidateSession(sessionId);

  if (removed) {
    return res.status(200).json({ message: 'Logged out successfully' });
  }

  return res.status(400).json({ error: 'Session not found' });
});

// Protected route example
app.get('/profile', authenticateToken, (req, res) => {
  const { sub: userId, username, sessionId } = req.user;
  const session = activeSessions.get(sessionId);

  return res.status(200).json({
    message: 'Profile retrieved successfully',
    user: {
      id: userId,
      username,
    },
    session: {
      sessionId,
      expiresAt: new Date(session.expiresAt).toISOString(),
      lastActivity: new Date(session.lastActivity).toISOString(),
    },
  });
});

// Active sessions overview (admin use only in production)
app.get('/auth/sessions', authenticateToken, (req, res) => {
  const userSessions = [];

  activeSessions.forEach((session, sessionId) => {
    if (session.userId === req.user.sub) {
      userSessions.push({
        sessionId,
        issuedAt: new Date(session.issuedAt).toISOString(),
        expiresAt: new Date(session.expiresAt).toISOString(),
        lastActivity: new Date(session.lastActivity).toISOString(),
      });
    }
  });

  return res.status(200).json({ sessions: userSessions });
});

// Clean up expired sessions periodically
setInterval(() => {
  const now = Date.now();
  activeSessions.forEach((session, sessionId) => {
    if (now > session.expiresAt) {
      activeSessions.delete(sessionId);
    }
  });
}, 15 * 60 * 1000); // Run every 15 minutes

const PORT = process.env.PORT || 3000;
app.listen(PORT, () => {
  console.log(`Server running on port ${PORT}`);
});

module.exports = {
  app,
  generateSessionToken,
  validateSessionToken,
  invalidateSession,
};